### PR TITLE
Prometheus: add support for new  _experimental_ sparse histograms/heatmaps

### PR DESCRIPTION
**Pull request by beorn7**
_Note: This is all experimental. So maybe this shouldn't be merged but can just stay here for reference while we play with it. I talked to @leeoniya about it already. @ryantxu, please sync with @leejones. This could be used as input for a parser to turn it into data frames. The goal is to hack something together for a demo at KubeCon (the final preparation of which would happen in the first week of May)._

See
https://docs.google.com/document/d/1Efu0LX-fgNWix6ehfeCR0FzeWtHvftWFNoy7cYW9nqU/edit#
for reference. This data is following example 4.

prom-matrix-histogram-no-labels.json is the result of a range query
for a single histogram (without any labels) over time. This could be
rendered as a single heatmap.

Query string:
`histogram_quantile(0.9, sum(rate(cortex_request_duration_seconds{status_code="200"}[5m])))`

prom-matrix-histogram-partitioned.json is the result of a range query
partitioned by certain labels. This results in many different
histograms, each with a number of samples over time (some without any
buckets, which means that all their buckets are empty). This could be
rendered as many different heatmaps.

Query string:
`histogram_quantile(0.9, sum by (route,cluster,job)(rate(cortex_request_duration_seconds{status_code="200"}[5m])))`

prom-vector-histogram-no-labels.json is the result of an instant
query. It's the same query string as the first above. Thus, this
contains a single histogram at a single point in time. This could be
rendered as a plain histogram as you find it in textbooks.

Signed-off-by: beorn7 <beorn@grafana.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by leejones** [comment]
> @ryantxu, please sync with @leejones.

@beorn7 I would help if I could, but I think maybe you meant @leeoniya? 😄

**Comment by beorn7** [comment]
@ryantxu thank you for promptly working on the parser. Does the format look OK? If you, I'd start to work on more efficient generation code within Prometheus. (The current one isn't using jsoniter yet. I just wanted to get it out quickly for initial testing.)

**Comment by ryantxu** [comment]
@beorn7 -- @leeoniya and I are now working on supporting this structure in the render layer.  

The one piece of information missing that would be helpful is the X bucket range.  Currently X has a single timestamp, but we have no indication what period it is aggregating over.  We can guess the interval by looking at adjacent values -- but that is pretty ugly and falls down when there is only one sample.

As I understand the prometheus format, it is: sparse Y, but still dense X.  When we convert to DataFrames, we are making it sparse in both X and Y.

**Comment by leeoniya** [comment]
> The one piece of information missing that would be helpful is the X bucket range.

@beorn7 this would be useful for other types of Prometheus responses, too. (we already try to do this by setting `timeField.config.interval`. a solution that allows the x aggregation interval to be exposed in the response would be great. not sure about the xLayout...we currently assume the returned x timestamps are `(..., timestamp]`

**Comment by beorn7** [comment]
tl;dr: The "X bucket range" should be equal to the value of the `step` parameter sent to the Prometheus server in the range query.


Longer answer:

With the current data and execution model, Prometheus cannot provide that information. Essentially, this information is part of the query, not part of the response.

In fact, you could write a query that isn't aggregating over a specific time frame at all (e.g. if you query the naked histogram, what you get is the counts since the last counter reset, which happened at an essentially unknown time in the past).

What you usually do is rate queries: `rate(my_histogram[10m])`, in which case each data point you return is an aggregation over the 10m prior to the timestamp.

For a heatmap, you would usually sync up the `step` parameter and the range in the rate, i.e. the above query should run with a step parameter of 10m, or vice versa, if your step is 1m, you should probably run `rate(my_histogram[1m])`, provided your scrape interval is short enough. This then leads up to the whole "rate_interval/xrate" discussion, i.e. to be precise, you should add one scrape interval (`rate(my_histogram[75s])` for a `step` of 1m and a scrape interval of 15s).

This is an old and hairy topic, and isn't specific to histograms or heatmaps. There is no easy and/or quick solution. I recommend to not tackle this issue at the same time as the heatmaps for the new Prometheus histograms.

I would think that it's safe to assume the X bucket range is equal to the `step` parameter in the Prometheus range query.

**Review by ryantxu — APPROVED** [review]
This will help the KubeCon demo... and has no effect otherwise

We should replace the test responses with smaller ones, but I think this is OK for now

File: public/app/plugins/panel/heatmap-new/fields.ts [hunk 1]
```diff
@@ -152,6 +160,26 @@ export const getExemplarsMapping = (heatmapData: HeatmapData, rawData: DataFrame
...
+  exemplars: DataFrame | undefined,
+  theme: GrafanaTheme2
+): HeatmapData => {
+  if (frame.meta?.type !== DataFrameType.ExperimentalSparseHistogram) {
+    return {
+      warning: 'Expected sparse heatmap format',
+      heatmap: frame,
+    };
+  }
+
```
Comment (ryantxu) [inline public/app/plugins/panel/heatmap-new/fields.ts:174]: @leeoniya -- I think https://github.com/grafana/grafana/pull/48564 will make this actually work, right?

**Review by leeoniya — APPROVED** [review]
